crates/blockchain/blockchain.rs

```diff
@@ -161,6 +143,120 @@ pub fn validate_block(
+/// - The finalized, safe and head hashes must correspond to already saved blocks.
+/// - The saved blocks should be in the correct order (finalized <= safe <= head).
+/// - They must be connected.
+///
+/// After the validity checks, the canonical chain is updated so that all head's ancestors
+/// and itself are made canonical.
+pub fn apply_fork_choice(
+    store: &Store,
+    head_hash: H256,
+    safe_hash: H256,
+    finalized_hash: H256,
+) -> Result<(), InvalidForkChoice> {
+    if head_hash.is_zero() {
+        return Err(InvalidForkChoice::InvalidHeadHash);
+    }
+
+    // We get the block bodies even if we only use headers them so we check that they are
+    // stored too.
+    let finalized_header_res = store.get_block_by_hash(finalized_hash).map_err(wrap)?;
+    let safe_header_res = store.get_block_by_hash(safe_hash).map_err(wrap)?;
+    let head_header_res = store.get_block_by_hash(head_hash).map_err(wrap)?;
+
+    // Check that we already have all the needed blocks stored and that we have the ancestors
+    // if we have the descendants, as we are working on the assumption that we only add block
+    // if they are connected to the canonical chain.
+    let (finalized, safe, head) = match (finalized_header_res, safe_header_res, head_header_res) {
+        (None, Some(_), _) => return Err(InvalidForkChoice::ElementNotFound),
+        (_, None, Some(_)) => return Err(InvalidForkChoice::ElementNotFound),
+        (Some(f), Some(s), Some(h)) => (f.header, s.header, h.header),
+        _ => return Err(InvalidForkChoice::Syncing),
+    };
+
+    // Check that we are not being pushed pre-merge
+    total_difficulty_check(&head_hash, &head, store)?;
+
+    // Check that the headers are in the correct order.
+    if finalized.number > safe.number || safe.number > head.number {
+        return Err(InvalidForkChoice::Unordered);
+    }
+
+    // If the head block is already in our canonical chain, the beacon client is
+    // probably resyncing. Ignore the update.
+    if is_canonical(store, head.number, head_hash).map_err(wrap)? {
+        return Ok(());
+    }
+
+    // Find out if blocks are correctly connected.
+    let Some(new_canonical_blocks) = find_link_with_canonical_chain(store, &head).map_err(wrap)?
+    else {
+        return Err(InvalidForkChoice::Disconnected(
+            error::ForkChoiceElement::Head,
+            error::ForkChoiceElement::Safe,
+        ));
+    };
+
+    let link_block_number = match new_canonical_blocks.last() {
+        Some((number, _)) => *number,
+        None => head.number,
+    };
+
+    // Check that finalized and safe blocks are either in the new canonical blocks, or already
+    // but prior to the canonical link to the new head. This is a relatively quick way of making
+    // sure that head, safe and finalized are connected.
+
+    if !(is_canonical(store, finalized.number, finalized_hash).map_err(wrap)?
```

Comment:
nit: I would add this checks _before_ finding the link with the canonical chain

Reply:
We can't really do that. The goal of this check is to see if the finalized and safe blocks are part of the new canonical chain. Only one of the checks can be done before, which is the case where finalized is exactly the same block as the head block. Same with safe. The other too checks, which are all other possible cases are divided between:

- They are already canonical and prior to the reorg point.
- They are in the new canonical blocks.

In both of this cases we need to know what the reorg point is so the link with the canonical chain must be found before.